### PR TITLE
feat: interp-engine support

### PR DESCRIPTION
**Pull request by hijohnnylin**
This adds support for [interp-engine](https://interp-engine.org) to circuit-tracer, which Neuronpedia is now fully switched to running all inference jobs on in production (including circuit tracing).

It mostly tries to not touch existing code, and adds some test coverage for the new engine, but it's still a somewhat significant change so I think we should be careful about reviewing/merging this.

No rush! Just making this public at least so that the Neuronpedia-used engine is open source. Happy to chat about this whenever you're free.

### CODE REVIEW

**Comment by hannamw** [comment]
thanks for opening this PR, Johnny - taking a look at this (and interp engine) now!

File: circuit_tracer/utils/demo_utils.py [hunk 1]
```diff
@@ -13,25 +13,21 @@
...
+    """Extract unembedding row vectors for the given token IDs.
 
     Args:
         model: A ``ReplacementModel`` instance.
-        token_ids: Vocabulary indices whose unembed columns to extract.
-        backend: ``"transformerlens"`` or ``"nnsight"``.
+        token_ids: Vocabulary indices whose unembed rows to extract.
+        backend: Ignored, and kept only so the demo notebook and its mirrored test keep calling
+            this unchanged. Every backend now exposes ``unembed_weight`` as ``(d_vocab, d_model)``,
+            so there is no longer an orientation to pick between.
```
Comment (hannamw) [inline circuit_tracer/utils/demo_utils.py:24]: If the backend is ignored, I'd prefer to remove it from the demo notebooks / tests

File: circuit_tracer/utils/demo_utils.py [hunk 2]
```diff
@@ -13,25 +13,21 @@
...
     Returns:
         List of 1-D tensors, one per token ID, each of shape ``(d_model,)``.
     """
-    unembed = model.unembed.W_U if backend == "transformerlens" else model.unembed_weight
-    d_vocab = model.tokenizer.vocab_size
-    if unembed.shape[0] == d_vocab:
-        return [unembed[tid] for tid in token_ids]
-    return [unembed[:, tid] for tid in token_ids]
+    unembed = model.unembed_weight
+    return [unembed[tid] for tid in token_ids]
```
Comment (hannamw) [inline circuit_tracer/utils/demo_utils.py:30]: almost not worth having a function for at this point! which is not a bad thing; now .unembed_weight is doing most of the work

File: circuit_tracer/utils/interventions.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+"""The intervention tuple, and the one rewriting of it that every backend needs.
+
+Both definitions below existed as byte-identical copies in the TransformerLens and nnsight
+backends. A third backend would have made a third copy of a rule about what user input means,
+which is the kind of thing that drifts silently.
```
Comment (hannamw) [inline circuit_tracer/utils/interventions.py:5]: Let's cut this

File: circuit_tracer/utils/interventions.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# "this feature, everywhere from here on" -- see `convert_open_ended_interventions`.
+Intervention = tuple[
+    int | torch.Tensor,
+    int | slice | torch.Tensor,
+    int | torch.Tensor,
+    int | float | torch.Tensor,
+]
+
+
+def convert_open_ended_interventions(
```
Comment (hannamw) [inline circuit_tracer/utils/interventions.py:22]: Creating a new interventions.py file for these made a lot of sense! It's cleaner this way

File: circuit_tracer/utils/tokenization.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
+"""Prompt -> token ids, shared by every backend.
+
+Attribution compares graphs across backends, so the backends have to agree on the token
+sequence before anything else can be compared. This lived as a byte-identical copy in each
+``ReplacementModel`` implementation; it is one function here so a third backend does not become
+a third copy that can drift.
```
Comment (hannamw) [inline circuit_tracer/utils/tokenization.py:6]: cut

File: circuit_tracer/replacement_model/replacement_model_transformerlens.py [hunk 6]
```diff
@@ -72,6 +65,21 @@ class TransformerLensReplacementModel(HookedTransformer):
     scan_name: str | list[str] | None
     backend: Literal["transformerlens"]
 
+    # The two weight accessors every backend shares, so that code holding a `ReplacementModel` can
+    # ask for the embedding matrices without first asking which backend it holds. Both are
+    # `[d_vocab, d_model]`, one row per token: that is what `embed_weight` already means on all
+    # three, and `unembed_weight` on the other two, so `W_U` is the odd one out and is transposed
+    # here rather than the convention bending around it. TransformerLens' own `W_E`/`W_U` stay
+    # exactly as they are for code that wants them in TL's terms.
```
Comment (hannamw) [inline circuit_tracer/replacement_model/replacement_model_transformerlens.py:73]: let's de-slop this

File: circuit_tracer/replacement_model/replacement_model.py [hunk 7]
```diff
@@ -19,8 +34,74 @@ class ReplacementModel:
...
+
+    Each factory is overloaded on ``backend`` so that a literal argument -- including the default --
+    yields the concrete backend class rather than a union of all three. Without this, reading a
+    backend-specific attribute off the result (``model.cfg``, which only TransformerLens has) is a
+    type error at every call site, since one member of the union is a plain ``nn.Module`` whose
+    ``__getattr__`` returns ``Tensor | Module``.
+
+    The embedding matrices are deliberately *not* backend-specific: every backend exposes
+    ``embed_weight`` and ``unembed_weight`` as ``(d_vocab, d_model)``, so code that only needs those
+    can stay generic. See ``tests/test_embedding_accessors.py``.
```
Comment (hannamw) [inline circuit_tracer/replacement_model/replacement_model.py:46]: cut

File: circuit_tracer/replacement_model/replacement_model.py [hunk 8]
```diff
@@ -19,8 +34,74 @@ class ReplacementModel:
 
     This class acts as a factory that creates the appropriate backend-specific ReplacementModel
     based on the backend parameter.
+
+    Each factory is overloaded on ``backend`` so that a literal argument -- including the default --
```
Comment (hannamw) [inline circuit_tracer/replacement_model/replacement_model.py:38]: nice, this seems like a useful change

File: circuit_tracer/replacement_model/replacement_model_nnsight.py [hunk 9]
```diff
@@ -58,6 +52,10 @@ class NNSightReplacementModel(LanguageModel):
     pre_logit_location: nn.Module  # type: ignore
     embed_loc: nn.Module
     unembed_loc: nn.Module
+    # Assigned in __init__ from the mapping config; declared here because they are part of the
+    # accessor every backend shares, `(d_vocab, d_model)` on all three.
```
Comment (hannamw) [inline circuit_tracer/replacement_model/replacement_model_nnsight.py:56]: feels a bit out of place

File: circuit_tracer/attribution/attribute_interp_engine.py [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    ctx = model.setup_attribution(input_ids)
+    activation_matrix = ctx.activation_matrix
+
+    logger.info(f"Precomputation completed in {time.time() - phase_start:.2f}s")
+    logger.info(f"Found {ctx.activation_matrix._nnz()} active features")
+
+    # A skip-connection transcoder is still needed during the forward pass, to compute the skip
+    # term the replacement model routes gradient through, so it can only be offloaded afterwards.
+    if offload and not model.skip_transcoder:
+        offload_handles += offload_modules(model.transcoders, offload)
```
Comment (hannamw) [inline circuit_tracer/attribution/attribute_interp_engine.py:137]: Does this need to be changed for the other backends too?

File: circuit_tracer/replacement_model/replacement_model_interp_engine.py [hunk 11]
```diff
@@ -0,0 +1,1554 @@
...
+            self._embed_tensor = tensor
+            return output
+
+        self._handles.append(embed.register_forward_hook(hook))
+
+    def _verify_freezes(self) -> None:
+        """Check on real tensors that both freezes do what they claim.
+
+        A short forward pass is cheap next to loading the model, and a freeze that silently does
+        not hold produces an attribution graph that looks entirely reasonable.
```
Comment (hannamw) [inline circuit_tracer/replacement_model/replacement_model_interp_engine.py:897]: should this be moved to a separate test? I'm trying to understand exactly which parts of freezing could break between different models, and which potential breakage points this verify function is checking.

This is a bit of a different approach compared to e.g. what I did for nnsight (where there's a list of models that are okay). It's nice that you don't have to fiddle with a model list, but feels a bit weird to have to do this every time.

File: circuit_tracer/replacement_model/replacement_model_interp_engine.py [hunk 12]
```diff
@@ -0,0 +1,1554 @@
...
+
+        error_vectors = feature_out - attribution_data["reconstruction"]
+        error_vectors[:, self.zero_positions] = 0
+        # Raw embedding rows, matching the NNSight backend. On Gemma these are sqrt(d_model) times
+        # smaller than TransformerLens reports, because TL folds that normalizer into `W_E` while HF
+        # applies it as a separate multiply after `embed_tokens`. The gradient is read on the
+        # pre-multiply tensor, so the factor cancels and token attributions still agree with TL to
+        # float precision; only this public vector differs, and it differs the same way NNSight's
+        # already does.
+        token_vectors = self.embed_weight[tokens].detach()
```
Comment (hannamw) [inline circuit_tracer/replacement_model/replacement_model_interp_engine.py:1060]: I wonder if it's worth standardizing this now that there will be 3 backends

File: tests/test_norm_kind.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
+"""Hermetic tests for `_subtracts_mean` (no model/GPU/network required, but interp-engine installed).
+
+The interp_engine backend introduces a transcoder's input tensor by substituting `x / scale` as a
+norm's input, and whether that scale is computed on a *centered* x is a property of the norm family.
+The test used to be the class name -- "rms" in `type(norm).__name__` -- which is wrong on the T5
+lineage: `T5LayerNorm` (T5, mT5, UMT5, Flan-T5) subtracts no mean despite what it is called.
```
Comment (hannamw) [inline tests/test_norm_kind.py:6]: I am a little confused about this - do we have transcoders for T5 models? Or is this in case we want to support them in the future?

**Review by hannamw — CHANGES_REQUESTED** [review]
Just did a preliminary review! I couldn't really review the replacement model or context in detail, and I am trusting that the tests do what they say. My thoughts in general:
- The changes to how ReplacementModel is instantiated are nice; it's good to not have 
- Moving the tokenization / intervention stuff to a separate file is also great, since there was a bunch of repeated code.
- Having a standardized (un)embed accessor is also handy
- In general, I'd like to keep the codebase readable by humans, so it would be nice to trim the comments in some places to make them a bit less slop-y. Some of the new comments are a bit hard to read
- The test coverage seems reasonably good, but I might want to standardize it across backends? That is, right now, for nnsight, we have a bunch of nnsight-only tests, and then some nnsight-transformerlens tests that ensure that they're the same. There are no interp-engine-only tests, and the tests ensuring that interp-engine is the same as nnsight/transformerlens cover different models than the nnsight-transformerlens tests. Maybe there's a good reason for this I'm not seeing though.

Anyway, this was just a first pass over the code - I haven't actually run it yet, but I'll try to do that soon. Would love to chat a bit about interp-engine and how it works! That might make it easier to properly review the replacement model / context

**Comment by hijohnnylin** [comment]
Hi Michael! Sorry for the delay on the followups here. I appreciate you reviewing it carefully. This integration was written by Opus, hence the obvious slop comments in many places, apologies for that.
I'll respond inline above now.

**Comment by hijohnnylin** [comment]
Edit: Will get to this tonight actually! Sorry somethign came up

File: circuit_tracer/utils/demo_utils.py (hunk 1 above)
Comment (hijohnnylin) [reply circuit_tracer/utils/demo_utils.py]: Removed

File: circuit_tracer/utils/demo_utils.py (hunk 2 above)
Comment (hijohnnylin) [reply circuit_tracer/utils/demo_utils.py]: get_unembed_vecs is now a one-line docstring and one return

File: circuit_tracer/utils/interventions.py (hunk 3 above)
Comment (hijohnnylin) [reply circuit_tracer/utils/interventions.py]: cut

File: circuit_tracer/utils/tokenization.py (hunk 5 above)
Comment (hijohnnylin) [reply circuit_tracer/utils/tokenization.py]: cut

File: circuit_tracer/replacement_model/replacement_model_transformerlens.py (hunk 6 above)
Comment (hijohnnylin) [reply circuit_tracer/replacement_model/replacement_model_transformerlens.py]: shortened, though can cut it completely if you'd prefer!

File: circuit_tracer/replacement_model/replacement_model.py (hunk 7 above)
Comment (hijohnnylin) [reply circuit_tracer/replacement_model/replacement_model.py]: cut

File: circuit_tracer/replacement_model/replacement_model_nnsight.py (hunk 9 above)
Comment (hijohnnylin) [reply circuit_tracer/replacement_model/replacement_model_nnsight.py]: cut

File: circuit_tracer/attribution/attribute_interp_engine.py (hunk 10 above)
Comment (hijohnnylin) [reply circuit_tracer/attribution/attribute_interp_engine.py:137]: Good call. Claude says nnsight had this guard, but that TL didn't. This means with skip-connection transcoders with offload to CPU (Llama 3.2-1B) fails on TL on `main`. Guard added to `attribute_transformerlens.py`

File: circuit_tracer/replacement_model/replacement_model_interp_engine.py (hunk 11 above)
Comment (hijohnnylin) [reply circuit_tracer/replacement_model/replacement_model_interp_engine.py]: Claude took your suggestion - it seems reasonable to me, but let me know you'd rather have different implementation.

> - VERIFIED_MODEL_TYPES = frozenset({"gemma2", "gemma3_text", "llama", "qwen3"}) in replacement_model_interp_engine.py. _configure_replacement_model now runs _verify_freezes only for architectures outside that set. Skipping is safe: the frozen-attention code reads the audit with a getattr(..., None) default, so the freezes themselves are unchanged.
> - New tests/test_interp_engine_freezes.py: builds a tiny untrained model per verified type (AutoConfig.for_model, 2 layers, d=64, gpt2 tokenizer, no gated downloads) and calls _verify_freezes() directly, plus one test that pins the set to those four so a new entry needs a case. It runs in the CPU build job: 5 tests pass on both CUDA and CPU, and the audit does real work (2 attention checks, 9 norm checks per model).
>

File: circuit_tracer/replacement_model/replacement_model_interp_engine.py (hunk 12 above)
Comment (hijohnnylin) [reply circuit_tracer/replacement_model/replacement_model_interp_engine.py]: Sorry that I'm not adding much value here personally, but seems Claude agreed and apparently found an issue with Gemma 3 on both nnsight and interp-engine. We're re-deploying Gemma 3 circuit tracing in Neuronpedia now with this patch.

> Standardized on: `token_vectors` is the embedding as it enters the residual stream, and the token gradient is read from that same tensor. TL already did this (`W_E` rows, `hook_embed`). nnsight and interp_engine now read layer 0's input instead of the embedding module's output.
> 
> This also fixed a real bug: Gemma-3 scales inside its embedding module, so both HF backends were reading the gradient post-scale while `token_vectors` was the raw row. On gemma-3-270m the interp_engine token→logit edges were exactly `1/sqrt(d_model)` (0.0395) of `grad(h0) · h0`; after the change the ratio is 1.0000, nnsight matches interp_engine on all 49,759 token edges, and on gemma-2-2b TL and interp_engine `token_vectors` are bit-identical.
> 
> The three nnsight steering tests now inject at layer 0's input; `test_token_vectors_differ_only_by_the_embedding_normalizer` became `test_token_vectors_agree`.
> 
> Verified: 10 nnsight tests pass (gemma-2-2b, gemma-3-1b, gemma-3-270m dummies, gpt-oss dummies), the ground-truth gemma case passes on nnsight (7.1 GiB) and interp_engine (7.3 GiB), ruff/pyright clean. TL ground-truth runs in CI; it OOMs on RAM locally.

I followed up by asking Claude to confirm impact on Gemma 3:

>  What was wrong on main is one part of the attribution graph: the edges out of the token (embedding) nodes, and only for Gemma-3, and only on the nnsight backend (the only HF-based backend main has). Those edges were 1/sqrt(d_model) too small: 1/34 on gemma-3-1b, 1/51 on gemma-3-4b.
> 
> Downstream effect: token nodes looked 34–51x less influential than they were, so pruning by influence would drop them or their edges, and the graph would overstate how much of a logit is explained by features versus by the prompt tokens directly. Every other edge was right.

File: tests/test_norm_kind.py (hunk 13 above)
Comment (hijohnnylin) [reply tests/test_norm_kind.py]: Removed - we test this already in interp-engine, no need to test here.

**Comment by hijohnnylin** [comment]
@hannamw
Responded to your suggestions individually. Please feel free to followup with additional requests/feedback or resolve if they seem ok.

Also added end-to-end verification- full graph generation tests for to the CI ("ground-truth" and "ground-truth-opt-in") which runs against real GPUs, comparing all 3 engines. The tests generate and compare graphs on:
gemma-2-2b (gemmascope 16k PLTs)
gemma-3-1b (gemmascope2-16k PLTs)
gemma-2-2b (CLTs)
qwen3-4b (PLTs)

These tests run on new PRs (not on every commit to a PR) and on merge to main and on manual trigger (I triggered it manually here).

Maybe it's worth adding a Llama model to CI too? Lmk what you think.

**Comment by hannamw** [comment]
Thanks Johnny! I will take a look at this after the ICLR deadline.
